src/Npgsql.NodaTime/Internal/TimestampConverters.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using NodaTime;
+using Npgsql.Internal;
+using static Npgsql.NodaTime.Internal.NodaTimeUtils;
+
+namespace Npgsql.NodaTime.Internal;
+
+public class InstantConverter : PgBufferedConverter<Instant>
+{
+    readonly bool _dateTimeInfinityConversions;
+
+    public InstantConverter(bool dateTimeInfinityConversions)
+        => _dateTimeInfinityConversions = dateTimeInfinityConversions;
+
+    public override bool CanConvert(DataFormat format, out BufferingRequirement bufferingRequirement)
+    {
+        bufferingRequirement = BufferingRequirement.FixedSize;
+        return base.CanConvert(format, out bufferingRequirement);
+    }
+
+    public override Size GetSize(SizeContext context, Instant value, ref object? writeState)
+        => sizeof(long);
+
+    protected override Instant ReadCore(PgReader reader)
+        => DecodeInstant(reader.ReadInt64(), _dateTimeInfinityConversions);
+
+    protected override void WriteCore(PgWriter writer, Instant value)
+        => writer.WriteInt64(EncodeInstant(value, _dateTimeInfinityConversions));
+}
+
+public class ZonedDateTimeConverter : PgBufferedConverter<ZonedDateTime>
+{
+    readonly bool _dateTimeInfinityConversions;
+
+    public ZonedDateTimeConverter(bool dateTimeInfinityConversions)
+        => _dateTimeInfinityConversions = dateTimeInfinityConversions;
+
+    public override bool CanConvert(DataFormat format, out BufferingRequirement bufferingRequirement)
+    {
+        bufferingRequirement = BufferingRequirement.FixedSize;
+        return base.CanConvert(format, out bufferingRequirement);
+    }
+
+    public override Size GetSize(SizeContext context, ZonedDateTime value, ref object? writeState)
+        => sizeof(long);
+
+    protected override ZonedDateTime ReadCore(PgReader reader)
+        => DecodeInstant(reader.ReadInt64(), _dateTimeInfinityConversions).InUtc();
+
+    protected override void WriteCore(PgWriter writer, ZonedDateTime value)
+    {
+        if (value.Zone != DateTimeZone.Utc && !LegacyTimestampBehavior)
```

Comment:
As discussed offline, we may want to go with the pass through resolver here. I'm slightly wary because of the perf implications just to add a validation check (though I did do so for DateTimeOffset). 

Alternatively we have to consider adding a real out-of-band validation mechanism to just PgTypeInfo (say a delegate, which array infos would automatically pick up and compose over), this could then be called during BindFormatAndLength.

Reply:
Will let you handle this one :)